Multinode functionality from the `multijoint` branch

kpamnany opened the pull request:
Recreated due to strange git problems with `multijoint`. Current state: single infer runs working. Also, fix for static compilation (can't pipe commands), and some affinity tweaking/testing.

Commit:
Multinode functionality from the `multijoint` branch

Recreated due to strange git problems with `multijoint`. Current state: single infer runs. Also, fix for static compilation (can't pipe commands), and some affinity tweaking/testing.

kpamnany commented:
Was the node/rank and thread ID display on the Log functions useful @jeff-regier? Should I revert the changes to that file?

jeff-regier commented:
Yes, it's somewhat useful.

On Wed, Mar 15, 2017, 9:02 AM kpamnany <notifications@github.com> wrote:

> Was the node/rank and thread ID display on the Log functions useful
> @jeff-regier <https://github.com/jeff-regier>? Should I revert the
> changes to that file?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jeff-regier/Celeste.jl/pull/620#issuecomment-286789777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIieC1JXF5_-B6qIJIXjdxhTvIzwdPZJks5rmAuRgaJpZM4MeGC6>
> .
>

jeff-regier commented on bin/infer-boxes.jl:
```diff
@@ -3,29 +3,40 @@
-        ns = [parse(Float64, x) for x in ss]
-        bb = BoundingBox(ns...)
+        lp = split(ln, '\t')
+        if length(lp) != 4
```
maybe print an error message too...failing silently could be a problem

jeff-regier commented on src/ParallelRun.jl:
```diff
@@ -136,21 +141,21 @@ function infer_init(rcfs::Vector{RunCamcolField},
-    timing.read_photoobj = toq()
-    Log.info("$(length(catalog)) primary sources")
+    timing.read_photoobj = timing.read_photoobj + toq()
+    #ntputs(grank(), threadid(), "$(length(catalog)) primary sources")
```
Are you worried the logging statements will print so much that they slow down the job? These print just once per box--not very often.

jeff-regier replied:
If you don't want the info commands printed, rather than comment these out, would you change Log.info to only print when the CELESTE_PROD environment variable is not set? I did that for Log.debug already.

jeff-regier replied:
Or if they aren't necessary any more because of your timing code, let's just delete them rather than commenting them out.

jeff-regier commented on src/ParallelRun.jl:
```diff
@@ -161,14 +166,14 @@ function infer_init(rcfs::Vector{RunCamcolField},
         tic()
         neighbor_map = Infer.find_neighbors(target_sources, catalog, images)
-        Log.info("neighbors found in $(toq()) seconds")
+        timing.find_neigh = timing.find_neigh + toq()
```
`+=`

jeff-regier commented on src/ParallelRun.jl:
```diff
@@ -250,7 +285,7 @@ function one_node_single_infer(config::Configs.Config,
             catch ex
                 if is_production_run || nthreads() > 1
-                    Log.exception(ex)
+                    Log.error(string(ex))
```
`Log.exception` gives quit a bit more information. Are you sure you want to change it?

jeff-regier commented on src/ParallelRun.jl:
```diff
@@ -377,33 +412,56 @@ save_results(outdir, box, results) =
+    times.write_results = toq()
 
-    puts_timing(timing)
+    puts_timing(times)
```
These instance of `InferTiming` are called `times` some places and `timing` elsewhere. Could we stick to just one of them?

jeff-regier commented on src/multinode_run.jl:
```diff
@@ -1,193 +1,700 @@
+                       ea_vec::Vector{ElboArgs},
+                       cfg_vec::Vector{Config{DEFAULT_CHUNK,Float64}};
+                       use_fft=false,
+                       trust_region=NewtonTrustRegion())
```
These `use_fft` and `trust_region` flags aren't used any more in `joint_infer.jl`. Can you call the code in `joint_infer` to initialize elbo args, either `initialize_elboargs_sources!` or `initialize_elboargs_sources_kernel!` rather than duplicating the code here?

jeff-regier commented on src/multinode_run.jl:
```diff
@@ -1,193 +1,700 @@
+                f = DeterministicVI.elbo
+            end
+            ntputs(grank(), threadid(), "maximizing source $s")
+            NewtonMaximize.maximize!(f, ea, cfg)
```
This call won't work because `maximize!` also needs `vp` based to it. Can you reuse the function in `joint_infer.jl` that calls NewtonMaximize, modifying that function as necessary? Those functions all are exercised by unit tests.

kpamnany commented:
Addressed all comments except the last two -- I'm still working on getting multi-node joint infer updated based on what's changed in `joint_infer.jl`. That'll be another PR.
